### PR TITLE
Pass in `req` and `res` directly to `PlumberStep$exec(req, res)`

### PR DESCRIPTION
**Pull request by schloerke**
* De-duplicates all matched formal arguments to avoid an error while processing the route
	* (No error is thrown anymore)
* Allow for unnamed arguments to be passed through to endpoint

PR task list:
- [x] Update NEWS
- [x] Add tests
- [x] Update documentation with `devtools::document()`

### CODE REVIEW

File: R/plumber-step.R [hunk 1]
```diff
@@ -56,21 +56,27 @@ PlumberStep <- R6Class(
...
       hookEnv <- new.env()
 
+      args <- c(
+        # add in `req`, `res` as they have been removed from `req$args`
+        list(req = req, res = res),
+        req$args
+      )
+
+      # str(args)
+
```
Comment (cpsievert) [inline R/plumber-step.R:70]:
```suggestion
```

File: NEWS.md [hunk 2]
```diff
@@ -12,12 +12,13 @@ plumber 1.0.0
 
 * An error will be thrown if multiple arguments are matched to an Plumber Endpoint route definition.
   While it is not required, it is safer to define routes to only use `req` and `res` when there is a possiblity to have multiple arguments match a single parameter name.
-  Use `req$argsPath`, `req$argsQuery`, and `req$argsBody` to access path, query, and postBody parameters respectively.
+  Use `req$argsQuery`, `req$argsPath`, and `req$argsBody` to access path, query, and body parameters respectively.
   See `system.file("plumber/17-arguments/plumber.R", package = "plumber")` to view an example with expected output and `plumb_api("plumber", "17-arguments")` to retrieve the api.
   (#637)
```
Comment (cpsievert) [inline NEWS.md:17]: Copy in the news that I sent you

File: R/plumber-step.R [hunk 3]
```diff
@@ -141,25 +137,33 @@ getRelevantArgs <- function(args, plumberExpression) {
...
-      "Can't call a Plumber function with duplicated matching formal arguments: ",
-      paste0(unique(matched_arg_names[duplicated_matched_arg_names]), collapse = ", "),
-      "\nPlumber recommends that the route's function signature be `function(req, res)`",
-      "\nand to access arguments via `req$args`, `req$argsPath`, `req$argsBody`, or `req$argsQuery`."
+      "Can't call a Plumber function with unnammed arguments. Missing names for argument(s) #",
+      paste0(unnamedArgs, collapse = ", "),
+      ". Names of argument list was: \"",
+      paste0(names(args), collapse = ","), "\""
     )
   }
```
Comment (cpsievert) [inline R/plumber-step.R:153]: Delete!!!

**Comment by schloerke** [comment]
@jcheng5 gentle ping

File: NEWS.md [hunk 4]
```diff
@@ -10,14 +10,21 @@ plumber 1.0.0
...
+  * The value supplied to `req` and `res` arguments in a route definition are now _always_ Plumber request and response objects. In the past, this was not guaranteed. (#666, #637)
+  * To assist with conflicts in argument names deriving from different locations, `req$argsQuery`, `req$argsPath`, and `req$argsBody` have been added to access query, path, and `req$body` parameters, respectively. For this reason, we suggest defining routes with only `req` and `res` (i.e., `function(req, res) {}`) and accessing argument(s) under these new fields to avoid naming conflicts. (#637)
+  * An error is no longer thrown if multiple arguments are matched to an Plumber Endpoint route definition. Instead, Plumber now retains the first named argument according to the highest priority match (`req$argsQuery` is 1st priority, then `req$argsPath`, then `req$argsBody`). (#666)
+  * Unnamed arguments that are added to `req$args` by filters or creating `req$argsBody` will no longer throw an error. They will only be passed through via `...` (#666)
```
Comment (jcheng5) [inline NEWS.md:19]: `s/arguments/elements/`

File: R/plumber-step.R [hunk 5]
```diff
@@ -56,21 +56,24 @@ PlumberStep <- R6Class(
...
     },
     #' @description step execution function
-    #' @param ... additional arguments for step execution
-    exec = function(...) {
-      allArgs <- list(...)
-      args <- getRelevantArgs(allArgs, plumberExpression=private$func)
-
+    #' @param req,res Request and response objects created by a Plumber request
+    exec = function(req, res) {
       hookEnv <- new.env()
```
Comment (jcheng5) [inline R/plumber-step.R:61]: Need a `parent` argument to `new.env()`

Comment (jcheng5) [reply R/plumber-step.R:61]: Possibilities: plumber (router) environment, emptyenv

Comment (jcheng5) [reply R/plumber-step.R:61]: emptyenv

Comment (jcheng5) [inline R/plumber-step.R:61]:
```suggestion
      hookEnv <- new.env(parent = emptyenv())
```

File: R/plumber-step.R [hunk 6]
```diff
@@ -56,21 +56,24 @@ PlumberStep <- R6Class(
...
-      hookEnv <- new.env()
+    #' @param req,res Request and response objects created by a Plumber request
+    exec = function(req, res) {
+      hookEnv <- new.env(parent = emptyenv())
 
+      args <- c(
+        # add in `req`, `res` as they have been removed from `req$args`
+        list(req = req, res = res),
+        req$args
+      )
```
Comment (jcheng5) [inline R/plumber-step.R:67]: This should be called right before execution in each of preexecStep, execStep, postexecStep

File: R/plumber-step.R [hunk 7]
```diff
@@ -102,21 +105,6 @@ PlumberStep <- R6Class(
...
-
-  if (length(unnamedArgs) > 0 ) {
-    stop("Can't call a Plumber function with unnammed arguments. Missing names for argument(s) #",
-         paste0(unnamedArgs, collapse=", "),
-         ". Names of argument list was: \"",
-         paste0(names(args), collapse=","), "\"")
-  }
-
   # Extract the names of the arguments this function supports.
   fargs <- names(formals(eval(plumberExpression)))
```
Comment (jcheng5) [inline R/plumber-step.R:109]: `eval()` can probably be removed. If `plumberExpression` is not a function at this point we already have problems.

Comment (jcheng5) [reply R/plumber-step.R:109]: If not removed, definitely needs an explicit evaluation environment.

File: R/plumber-step.R [hunk 8]
```diff
@@ -125,6 +113,12 @@ getRelevantArgs <- function(args, plumberExpression) {
...
 
+  # fast return
+  # also works with unnamed arguments
+  if (identical(fargs, "...")) {
+    return(args)
+  }
+
   # If only req and res are found in function definition...
   # Only call using the first matches of req and res.
   #   This allows for body content to have `req` and `res` named arguments and not duplicated values which cause issues.
```
Comment (jcheng5) [inline R/plumber-step.R:124]:
```suggestion
```

File: R/plumber.R [hunk 9]
```diff
@@ -614,18 +601,20 @@ Plumber <- R6Class(
...
-            # req, res
-            # query string params and any other `req$args`
-            ## Query string params have been added to `req$args`.
-            ## At this point, can not include both `req,res` and `req$argsQuery`. So using `req$args`
+            # (does not contain req or res)
+            # will contain all args added in filters
+            # `req$argsQuery` is available, but already absorbed into `req$args`
             req$args,
-            # path params
+            # path is more important than query
```
Comment (jcheng5) [inline R/plumber.R:608]:
```suggestion
            # path is more important than body
```

File: R/plumber.R [hunk 10]
```diff
@@ -614,18 +601,20 @@ Plumber <- R6Class(
...
-            # path params
+            # path is more important than body
             req$argsPath,
-            # body params
+            # body is added last
             req$argsBody
           )
 
-          return(do.call(h$exec, req$args))
+          # req$args will be filled inside `h$exec()`
```
Comment (jcheng5) [inline R/plumber.R:614]:
```suggestion
```

**Review by jcheng5 — APPROVED** [review]

**Pull request merged**